pkg/metrics/metrics.go

```diff
@@ -0,0 +1,147 @@
+func (m Counter) Increment() {
+	m.prometheusCounter.Inc()
+}
+
+// ConditionalIncrement increments the Counter if the condition is true
+func (m Counter) ConditionalIncrement(condition bool) {
+	if condition {
+		m.prometheusCounter.Inc()
+	}
+}
+
+// Gauge is a wrapper of the prometheus gauge type
+type Gauge struct {
+	Name            string
+	Help            string
+	prometheusGauge prometheus.Gauge
+}
+
+// Collector returns a collector for the metric
+func (m Gauge) Collector() prometheus.Collector {
+	return m.prometheusGauge
+}
+
+// Set the gauge if it has been initialized
+func (m Gauge) Set(value float64) {
+	m.prometheusGauge.Set(value)
+}
+
+// counters
+var (
+	// UserSignupUniqueTotal should be incremented only the first time a user signup is created, there should be 1 for each unique user
+	UserSignupUniqueTotal = initCounter("user_signups_total", "Total number of unique User Signups")
+
+	// UserSignupProvisionedTotal should be incremented each time a user signup is provisioned, can be multiple times per user if they reactivate multiple times
+	UserSignupProvisionedTotal = initCounter("user_signups_provisioned_total", "Total number of Provisioned User Signups")
+
+	// UserSignupBannedTotal should be incremented each time a user signup is banned
+	UserSignupBannedTotal = initCounter("user_signups_banned_total", "Total number of Banned User Signups")
+
+	// UserSignupDeactivatedTotal should be incremented each time a user signup is deactivated, can be multiple times per user if they reactivate multiple times
+	UserSignupDeactivatedTotal = initCounter("user_signups_deactivated_total", "Total number of Deactivated User Signups")
+
+	// UserSignupAutoDeactivatedTotal should be incremented each time a user signup is automatically deactivated, can be multiple times per user if they reactivate multiple times
+	UserSignupAutoDeactivatedTotal = initCounter("user_signups_auto_deactivated_total", "Total number of Automatically Deactivated User Signups")
+)
+
+// gauges
+var (
+	// MasterUserRecordGauge should reflect the current number of master user records in the system
+	MasterUserRecordGauge = initGauge("master_user_record_current", "Current number of Master User Records")
+)
+
+// collections
+var (
+	allCounters = []*Counter{}
+	allGauges   = []*Gauge{}
+)
+
+func initCounter(name, help string) *Counter {
+	c := prometheus.NewCounter(prometheus.CounterOpts{
+		Name: name,
+		Help: help,
+	})
+	m := &Counter{
+		Name:              name,
+		Help:              help,
+		prometheusCounter: c,
+	}
+	allCounters = append(allCounters, m)
+	return m
+}
+
+func initGauge(name, help string) *Gauge {
+	g := prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: name,
+		Help: help,
+	})
+	m := &Gauge{
+		Name:            name,
+		Help:            help,
+		prometheusGauge: g,
+	}
+	allGauges = append(allGauges, m)
+	return m
+}
+
+// RegisterCustomMetrics registers the custom metrics
+func RegisterCustomMetrics() {
```

Comment:
I haven't seen that you would call this function in any of the tests - shouldn't we verify it too?

Reply:
I wasn't sure if this would be more like testing the prometheus API than our code but sure I can try to add some tests for this package in general

Reply:
Tests added in https://github.com/codeready-toolchain/host-operator/pull/309/commits/6a4ed48ecc4e0fd29f67e3de9a13d94a41c31664